pallets/connectors/src/lib.rs

```diff
@@ -0,0 +1,507 @@
+			// Depends on https://github.com/centrifuge/centrifuge-chain/issues/842
+			//
+			// TODO: only allow calling add_tranche when
+			// both the name and symbol are non-zero values.
+			Self::do_send_message(
+				who,
+				Message::AddTranche {
+					pool_id,
+					tranche_id,
+					token_name: [0; 32],
+					token_symbol: [0; 32],
+				},
+				domain,
+			)?;
+
+			Ok(())
+		}
+
+		/// Update a token price
+		#[pallet::weight(< T as Config >::WeightInfo::update_token_price())]
+		pub fn update_token_price(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			tranche_id: TrancheIdOf<T>,
+			domain: Domain,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+
+			ensure!(
+				T::PoolInspect::tranche_exists(pool_id, tranche_id),
+				Error::<T>::TrancheNotFound
+			);
+
+			let latest_price = T::PoolInspect::get_tranche_token_price(pool_id, tranche_id)
+				.ok_or(Error::<T>::MissingPrice)?;
+
+			Self::do_send_message(
+				who,
+				Message::UpdateTokenPrice {
+					pool_id,
+					tranche_id,
+					price: latest_price.price,
+				},
+				domain,
+			)?;
+
+			Ok(())
+		}
+
+		/// Update a member
+		#[pallet::weight(< T as Config >::WeightInfo::update_member())]
+		pub fn update_member(
+			origin: OriginFor<T>,
+			address: DomainAddress<Domain>,
+			pool_id: PoolIdOf<T>,
+			tranche_id: TrancheIdOf<T>,
+			valid_until: Moment,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+
+			// Check that the origin is a member of this tranche token or
+			// is a MemberListAdmin and thus allowed to add other members.
+			ensure!(
+				T::Permission::has(
+					PermissionScope::Pool(pool_id),
+					who.clone(),
+					Role::PoolRole(PoolRole::TrancheInvestor(tranche_id, Self::now()))
+				) || T::Permission::has(
+					PermissionScope::Pool(pool_id),
+					who.clone(),
+					Role::PoolRole(PoolRole::MemberListAdmin)
+				),
+				BadOrigin
+			);
+
+			T::Permission::add(
+				PermissionScope::Pool(pool_id),
+				address.into_account_truncating(),
+				Role::PoolRole(PoolRole::TrancheInvestor(tranche_id, valid_until)),
+			)?;
+
+			Self::do_send_message(
+				who,
+				Message::UpdateMember {
+					pool_id,
+					tranche_id,
+					valid_until,
+					address: address.address,
+				},
+				address.domain,
+			)?;
+
+			Ok(())
+		}
+
+		/// Transfer tranche tokens to a given address
+		#[pallet::weight(< T as Config >::WeightInfo::transfer())]
+		pub fn transfer(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			tranche_id: TrancheIdOf<T>,
+			address: DomainAddress<Domain>,
+			amount: <T as pallet::Config>::Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+
+			// Check that the destination is a member of this tranche token
+			ensure!(
+				T::Permission::has(
+					PermissionScope::Pool(pool_id),
+					address.into_account_truncating(),
+					Role::PoolRole(PoolRole::TrancheInvestor(tranche_id, Self::now()))
+				),
+				BadOrigin
```

Comment:
I think we should add better error resolution to the permissions trait. The `BadOrigin` is getting really opaque if we have batched calls and so on. 

This is on my plate but I would appreciate some input on what you think is a good approach.

Reply:
Good point on the batch scenario. We can introduce a `InvalidTransferPermissions` error variant for this case at least for the time being. What do you think?